cpmpy/solvers/z3.py

```diff
@@ -0,0 +1,303 @@
+
+        return has_sol
+
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        import z3
+
+        if is_num(cpm_var): # shortcut, eases posting constraints
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return z3.Not(self.solver_var(cpm_var._bv))
+
+        # create if it does not exit
+        if cpm_var not in self._varmap:
+            # we assume al variables are user variables (because nested expressions)
+            self.user_vars.add(cpm_var)
+            if isinstance(cpm_var, _BoolVarImpl):
+                revar = z3.Bool(str(cpm_var))
+            elif isinstance(cpm_var, _IntVarImpl):
+                revar = z3.Int(str(cpm_var))
+                # set bounds
+                self.z3_solver.add(revar >= cpm_var.lb)
+                self.z3_solver.add(revar <= cpm_var.ub)
+            else:
+                raise NotImplementedError("Not a know var {}".format(cpm_var))
+            self._varmap[cpm_var] = revar
+
+        return self._varmap[cpm_var]
+
+
+    # if TEMPLATE does not support objective functions, you can delete objective()/_make_numexpr()
+    def objective(self, expr, minimize=True):
+        """
+            Post the given expression to the solver as objective to minimize/maximize
+
+            'objective()' can be called multiple times, only the last one is stored
+
+            (technical side note: any constraints created during conversion of the objective
+            are premanently posted to the solver)
+        """
+        # objective can be a nested expression for z3
+        obj = self._z3_expr(expr)
+        if minimize:
+            TEMPLATEpy.Minimize(obj)
+        else:
+            TEMPLATEpy.Maximize(obj)
+
+    def __add__(self, cpm_con):
+        """
+        Post a (list of) CPMpy constraints(=expressions) to the solver
+
+        Note that we don't store the constraints in a cpm_model,
+        we first transform the constraints into primitive constraints,
+        then post those primitive constraints directly to the native solver
+
+        :param cpm_con CPMpy constraint, or list thereof
+        :type cpm_con (list of) Expression(s)
+        """
+        # Z3 supports nested expressions,
+        # so we recursively translate our expressions to theirs
+
+        # that also means we don't need to extract user variables here
+        # we store them directly in `solver_var()` itself.
+        #self.user_vars.update(get_variables(cpm_con))
+
+        # only complication is that a list is implicitly an 'And' for us,
+        # as well as being a list of arguments, so top-level lists first
+        if is_any_list(cpm_con):
+            # recursively
+            for con in cpm_con:
+                self.__add__(con)
+        else:
```

Comment:
This isn't really compatible with our ideas for putting __add__ in the superclass and supporting by supporting a `transform()` function. 
I think the logic for posting the constraint to the z3 solver should still be contained in the `_post_constraint` method as is done in other solvers.

Reply:
I see, I agree. So yes, no transforms here and so this content would become post_constraint, go ahead

Reply:
For future reference: the transform() function of z3 should then just make sure it returns a list of constraints